proposals/open-telemetry/microprofile-opentelemetry-tracing.asciidoc

```diff
@@ -0,0 +1,193 @@
+
+=== Automatic Instrumentation
+
+Jakarta RESTful Web Services (server and client), and MicroProfile REST Clients are automatically enlisted to
+participate in distributed tracing without code modification.
+
+=== Manual Instrumentation
+
+Explicit manual instrumentation can be added into a MicroProfile application in the following ways:
+
+==== @WithSpan
+
+Annotating a method in any Jakarta CDI aware beans with the `io.opentelemetry.extension.annotations.WithSpan`
+annotation. This will create a new Span and establish any required relationships with the current Trace context.
+
+Method parameters can be annotated with the `io.opentelemetry.extension.annotations.SpanAttribute` annotation to
+indicate which method parameters should be part of the Trace.
+
+Example:
+[source,java]
+----
+@ApplicationScoped
+class SpanBean {
+    @WithSpan
+    void span() {
+
+    }
+
+    @WithSpan("name")
+    void spanName() {
+
+    }
+
+    @WithSpan(kind = SERVER)
+    void spanKind() {
+
+    }
+
+    @WithSpan
+    void spanArgs(@SpanAttribute(value = "arg") String arg) {
+
+    }
+}
+----
+
+==== Obtain a SpanBuilder
+
+By obtaining a `SpanBuilder` from the current `Tracer` and calling
+`io.opentelemetry.api.trace.Tracer.spanBuilder(String)`. In this case, it is the developer's responsibility to ensure
+that the `Span` is properly created, closed, and propagated.
+
+Example:
+[source,java]
+----
+@RequestScoped
+@Path("/")
+public class SpanResource {
+    @Inject
+    Tracer tracer;
+
+    @GET
+    @Path("/span/new")
+    public Response spanNew() {
+        Span span = tracer.spanBuilder("span.new")
+                .setSpanKind(INTERNAL)
+                .setParent(Context.current().with(this.span))
+                .setAttribute("my.attribute", "value")
+                .startSpan();
+
+        span.end();
+
+        return Response.ok().build();
+    }
+}
+----
+
+NOTE: Start and end a new `Span` will add a child `Span` to the current one enlisted by the automatic instrumentation
+of Jakarta REST Applications.
+
+==== Obtain the current Span
+
+By obtaining the current `Span` to add attributes. The Span lifecycle is managed by the implementation.
+
+Example:
+[source,java]
+----
+@RequestScoped
+@Path("/")
+public class SpanResource {
+    @Inject
+    Span span;
+
+    @GET
+    @Path("/span/current")
+    public Response spanCurrent() {
+        span.setAttribute("my.attribute", "value");
+        return Response.ok().build();
+    }
+}
+----
+
+=== Agent Instrumentation
+
+TODO
+
+=== Access to the OpenTelemetry Tracing API
+
+The implemention of MicroProfile OpenTelemetry must provide the following CDI beans for supporting contextual instance injection:
+
+- `io.opentelemetry.api.OpenTelemetry`
+- `io.opentelemetry.api.baggage.Baggage`
+- `io.opentelemetry.api.trace.Span`
+- `io.opentelemetry.api.trace.Tracer`
+
+=== Configuration
+
+OpenTelemetry must be configured by MicroProfile Config following the configuration properties detailed in:
```

Comment:
We should list the properties from opentelemetry that we must support in the integration with MP Config and make sure we have tests for that. To make sure the bridge between MP Config and the underlying OT infra.  
We should add a TODO for that.

Reply:
My idea was just to support everything detailed in the OTel docs, with the option to skip specific configurations related to an exporter. For instance, if the implementation does not support the Jaeger Exporter for some reason, then of course the related Jaeger configuration is also not supported.

Reply:
Ok, but then we need to lock one version of OTel and implement all the property mappings with MP Config, no?

Reply:
We need to lock down which version of OTel we are aligning with and provide the relevant MP Config properties for the configurations we plan to do. I think adding todo at this stage might be a better way forward.

Reply:
This will be locked down of course. We need to use the OTel API for the TCK, which is tied to a version and the docs and specs are also tied to a version. On the link, instead of referring to `main` we just refer to the version we use.

Reply:
sounds good

Reply:
Lockdown OTel version and updated links.